tests/scanner/test_file_downloader.py

```diff
@@ -0,0 +1,171 @@
+
+from matrix_content_scanner.utils.errors import ContentScannerRestError
+from tests.testutils import (
+    MEDIA_PATH,
+    SMALL_PNG,
+    get_base_media_headers,
+    get_content_scanner,
+)
+
+
+class FileDownloaderTestCase(aiounittest.AsyncTestCase):
+    def setUp(self) -> None:
+        # Set a fixed base URL so that .well-known discovery doesn't get in the way.
+        content_scanner = get_content_scanner(
+            {"download": {"base_homeserver_url": "http://my-site.com"}}
+        )
+        self.downloader = content_scanner.file_downloader
+
+        self.media_status = 200
+        self.media_body = SMALL_PNG
+        self.media_headers = get_base_media_headers()
+
+        async def _get(url: str) -> Tuple[int, bytes, Headers]:
+            """Mock for the _get method on the file downloader that doesn't serve a
+            .well-known client file.
+            """
+            if (
+                url.endswith("/_matrix/media/v3/download/" + MEDIA_PATH)
+                or "/_matrix/media/v3/thumbnail/" + MEDIA_PATH in url
+                or url.endswith("/_matrix/media/r0/download/" + MEDIA_PATH)
+                or "/_matrix/media/r0/thumbnail/" + MEDIA_PATH in url
+            ):
+                return self.media_status, self.media_body, self.media_headers
+            elif url.endswith("/.well-known/matrix/client"):
+                return 404, b"Not found", Headers()
+
+            raise RuntimeError("Unexpected request on %s" % url)
+
+        # Mock _get so we don't actually try to download files.
+        self.get_mock = Mock(side_effect=_get)
+        self.downloader._get = self.get_mock  # type: ignore[assignment]
+
+    async def test_download(self) -> None:
+        """Tests that downloading a file works."""
+        media = await self.downloader.download_file(MEDIA_PATH)
+        self.assertEqual(media.content, SMALL_PNG)
+        self.assertEqual(media.content_type, "image/png")
+
+        # Check that we tried downloading from the set base URL.
+        args = self.get_mock.call_args
+        self.assertTrue(args[0][0].startswith("http://my-site.com/"))
+
+    async def test_no_base_url(self) -> None:
+        """Tests that configuring a base homeserver URL means files are downloaded from
+        that homeserver (rather than the one the files were uploaded to) and .well-known
+        discovery is bypassed.
+        """
+        self.downloader._base_url = None
+        await self.downloader.download_file(MEDIA_PATH)
+
+        # # Check that we've tried making a .well-known discovery request before
+        # # downloading the file.
+        # self.assertEqual(self.get_mock.call_count, 1)
+        # self.assertEqual(
+        #     self.get_mock.mock_calls[0], call("https://foo/.well-known/matrix/client")
+        # )
+        # self.assertEqual(
+        #     self.get_mock.mock_calls[1],
+        #     call("https://foo/_matrix/media/v3/download/" + MEDIA_PATH),
+        # )
```

Comment:
This will be uncommented when well-known discovery support is added in a future PR.